src/com/ms/silverking/net/async/AsyncBase.java

```diff
@@ -263,160 +258,180 @@ public void doWork(Connection connection) {
-        connection = connectionCreator.createConnection(channel, selectorController, 
-                                                        listener, workPool, debug);
-
-        if (authResult.isSuccessful()) {
-            Log.info("Authenticator: authId["+authResult.getAuthId().get()+"] is obtained in ["
-                    + (serverside ? "ServerSide" : "ClientSide")+ "] by ["+authenticator.getName()+"]");
-            connection.setAuthResult(authResult);
-        }
-        connection.start();
-        if (Connection.statsEnabled) {
-            connections.add(connection);
-        }
-        return connection;
+    String connInfo = channel.socket() != null ? channel.socket().toString() : "nullSock";
+    AuthResult authResult = authenticator.syncAuthenticate(channel.socket(), serverside,
+        defAuthenticationTimeoutInMillisecond);
+    if (authResult.isFailed()) {
+      switch (authResult.getFailedAction()) {
+      case GO_WITHOUT_AUTH:
+        break;
+      case THROW_ERROR:
+        String msg = "Connection " + connInfo + " fails to be authenticated from " + (serverside ?
+            "ServerSide" :
+            "ClientSide");
+        throw authResult.getFailCause().isPresent() ? new AuthFailure(msg,
+            authResult.getFailCause().get()) : new AuthFailure(msg);
+      case ABSORB:
+        throw new ConnectionAbsorbException(channel, connInfo, listener, serverside,
+            authResult.getFailCause().orElse(null));
+      default:
+        throw new RuntimeException("Connection " + connInfo + " fails to be authenticated from " + (serverside ?
+            "ServerSide" :
+            "ClientSide" + " and action for this failure has NOT been defined: " + "please check the behaviour of " +
+                "injected authenticator [" + authenticator.getName() + "]"));
+      }
     }
-    
-    /**
-     * Assigned a new ServerSocketChannel to a selector and adds it to that
-     * selector. Also, configures the channel as non-blocking.
-     * @param serverChannel
-     * @throws IOException
-     */
-    protected void addServerChannel(ServerSocketChannel serverChannel) throws IOException {
-        SelectorController<T>    selectorController;
-
-        serverChannel.configureBlocking(false);
-        selectorController = cscAssigner.assignChannelToSelectorController(serverChannel, selectorControllers);
-        selectorController.addServerChannel(serverChannel);
+
+    if (logConnections) {
+      Log.warningf("AsyncBase addConnection: %s sBuf %d rBuf %d", channel, channel.socket().getSendBufferSize(), channel.socket().getReceiveBufferSize());
     }
-    
-    public void writeStats() {
-        for (Connection c : connections) {
-            try {
-                connectionStatsWriter.writeStats(c);
-            } catch (Exception e) {
-                e.printStackTrace();
-            }
-        }
+
+    try {
+      channel.configureBlocking(false);
+    } catch (IOException ioe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)